platforms/avr/drivers/ws2812_spi_int.c

```diff
@@ -0,0 +1,96 @@
+                 "        ldi  r25, 0xE0\n\t"           //     r25 = 0xE0;          11100000 -> zero bit (375ns up)
+                 "        sbrc r24, 7\n\t"              // else
+                 "        ldi  r25, 0xFE\n\t"           //     r25 = 0xFE;          11111110 -> one bit (875ns up)
+                 "        add  r24, r24\n\t"            // *Z <<= 2;                rotate the current byte one bit to the left
+                 "        st   Z+, r24\n\t"             // Z++;                     store it in memory and point to the next byte (the incremented value won't be
+                 "        lds  r24, remaining_bits\n\t" //                          stored unless there are no more bits, so this saves one instruction)
+                 "        subi r24, 1\n\t"              // remaining_bits--;
+                 "        breq no_more_bits\n\t"        // if remaining_bits == 0   are there more bits to send?
+                 "        sts  remaining_bits, r24\n\t" //     goto no_more_bits    yes -> store the new number of bits and continue
+                 "exit_isr:\n\t"
+                 "        pop  r31\n\t" //                                          restore the registers
+                 "        pop  r30\n\t"
+                 "        pop  r24\n\t"
+                 "        out  0x2E, r25\n\t" //           SPDR = r25               and at the very last moment, send the new bit. That way, the main code will
+                 "        pop  r25\n\t"       //                                    have the maximum possible time to do other things before the next interrupt.
+                 "        out  0x3F, r25\n\t"
+                 "        pop  r25\n\t"
+                 "        reti\n\t"
+                 "no_more_bits:\n\t"
+                 "        ldi  r24, 8\n\t"              //                          no more bits to send from the current byte; so we will send 8 more
+                 "        sts  remaining_bits, r24\n\t" // remaining_bits = 8;
+                 "        sts  led_data, r30\n\t"       // led_data = Z;            store the pointer to the next byte
+                 "        sts  led_data+1, r31\n\t"     //                          (it was incremented previously, so we save one instruction)
+                 "        lds  r24, nbytes\n\t"
+                 "        subi r24, 1\n\t"
+                 "        sts  nbytes, r24\n\t" //         nbytes--;                one byte less to send
+                 "        brne exit_isr\n\t"    //         if nbytes == 0           if there are more bytes to send, just exit the ISR
+                 "        in   r24, 0x2C\n\t"   //             goto exit_isr
+                 "        andi r24, 0x7F\n\t"
+                 "        out  0x2C, r24\n\t" //           SPCR &= ~(1<<SPIE)       but if there aren't more, disable the interrupts
+                 "        rjmp exit_isr\n\t");
+}
```

Comment:
This code would take 8 CPU cycles from the SPDR write to the return to the interrupted code, then after 8 more cycles the SPI interrupt would be activated again. After the SPI interrupt we would have:
- at least 5 cycles of interrupt response time;
- 3 cycles for the `rjmp` command in the interrupt vector slot;
- 36 cycles in the best case to the completion of the `out` instruction which writes to SPDR.

So the total bit time would be at least 16 + 5 + 3 + 36 = 60 CPU cycles, or 3750 ns (266 kbit/s); with that speed you would be able to send the data for just 11 LEDs in 1 ms.  And any extra interrupt latency may push the time above the critical 9000 ns, which may cause a premature reset and mess up the LED state.

Reply:
To update the LEDs 60 times per second you have to update them once every 16,666ms. In that time you can update 183 LEDs... Do the LED routines update the LEDs faster than that? Mainly because it makes no sense...

Reply:
The RGB Matrix code has a setting to throttle updates, so it won't call the driver more than once in 16 ms with the default configuration (it may call it less frequently though if the MCU cannot keep up with things).  The RGBLIGHT code, however, does not have anything like that, and things like switching rgblight layers may cause it to emit back-to-back driver calls (#18338 fixed that part of code to at least buffer the updates until the next pass of the QMK main loop, but further state changes may make the code call the driver more frequently than once every 16 ms).

Reply:
Also a major issue here is that while you are slowly pushing the WS2812 data out, the main QMK code runs at a snail pace (8 cycles for the main code, then at least 52 cycles (probably even more after recent changes) for the interrupt handler… that's 13% of the original performance).  I'm not sure that such behavior would actually be useful.